forgerock-auth/src/main/java/org/forgerock/android/auth/OAuth2ResponseHandler.java

```diff
@@ -33,20 +35,30 @@ class OAuth2ResponseHandler implements ResponseHandler {
      * @param response The response from /authorize endpoint
      * @param listener Listener for receiving OAuth APIs related changes
      */
-    void handleAuthorizeResponse(Response response, FRListener<String> listener) {
-        if (response.isRedirect()) {
-            String location = response.header("Location");
-            Uri redirect = Uri.parse(location);
-            String code = redirect.getQueryParameter("code");
-            if (code != null) {
-                Listener.onSuccess(listener, code);
+    void handleAuthorizeResponse(Response response, @NonNull String state, FRListener<String> listener) {
+        try {
+            if (response.isRedirect()) {
+                String location = response.header("Location");
+                Uri redirect = Uri.parse(location);
+                String code = redirect.getQueryParameter("code");
+                String responseState = redirect.getQueryParameter("state");
+                if (responseState == null || !responseState.equals(state)) {
```

Comment:
@witrisna so basically we are sending a state and expecting a same state back from server ?

Reply:
Correct, since we send the state to server, according to the spec the server has to return it and match with what we sent.